src/6model/reprs/Semaphore.c

```diff
@@ -113,25 +160,60 @@ static const MVMREPROps Semaphore_this_repr = {
     NULL, /* describe_refs */
 };
 
-MVMint64 MVM_semaphore_tryacquire(MVMThreadContext *tc, MVMSemaphore *sem) {
-    int r;
+MVMuint64 MVM_semaphore_tryacquire(MVMThreadContext *tc, MVMSemaphore *sem) {
+    MVMuint64 count, minus;
     MVM_telemetry_timestamp(tc, "Semaphore.tryAcquire");
-    r = uv_sem_trywait(sem->body.sem);
-    return !r;
+#ifdef MVM_USE_C11_ATOMICS
+    count = atomic_load_explicit(&sem->body.count, memory_order_acquire);
+    if ((minus = count))
+        atomic_store_explicit(&sem->body.count, minus -= 1, memory_order_release);
+#else
+    MVM_barrier();
+    count = MVM_load(&sem->body.count);
+    MVM_store(count, minus = count && count - 1);
+    MVM_barrier();
+#endif
+    return minus < count;
 }
 
 void MVM_semaphore_acquire(MVMThreadContext *tc, MVMSemaphore *sem) {
     unsigned int interval_id;
     interval_id = MVM_telemetry_interval_start(tc, "Semaphore.acquire");
     MVMROOT(tc, sem, {
+        MVMuint64 count;
         MVM_gc_mark_thread_blocked(tc);
-        uv_sem_wait(sem->body.sem);
+#ifdef MVM_USE_C11_ATOMICS
+        while (atomic_flag_test_and_set_explicit(&sem->body.waits, memory_order_acquire))
+            MVM_thread_yield(tc);
+        while (!(count = atomic_load_explicit(&sem->body.count, memory_order_acquire)))
+            MVM_thread_yield(tc);
```

Comment:
I did this in the MoarVM GC, before replacing it with condition variables. In theory it should be fine; in practice it wasn't, because yield can result on the thread being put back on the CPU in very short order, and so it's not that much better than busy-waiting.

Reply:
Do you think `uv_rwlock_t` might apply here?